text/0000-rand-crate-redesign.md

```diff
@@ -0,0 +1,991 @@
+useful for security reviews.
+
+**Question:**
+If `rand`/`Rng` is split into multiple parts, several approaches are possible;
+what to do here is very much undecided.
+
+First, "strong RNGs" designed for cryptographic usage could:
+
+1.  Use a dedicated `CryptoRng` trait, indepedant of `Rng`, with methods
+    returning a `Result` on failure
+2.  As (1), but add a wrapper struct implementing `Rng` for a `CryptoRng` which
+    panics on errors
+3.  Use the standard `Rng` trait and be designed such that the only time
+    failure is possible is during creation (`fn new() -> Result<Self>`)
+4.  Do not include any generators which may fail in `rand` (leave to dedicated
+    crypto libraries, if at all) [practically, this is equivalent to 3 aside
+    from choice of generators to include]
+
+Second, the `rand` crate could:
+
+1.  Remain a single crate
+2.  Have a sub-crate `os_rand` covering only the `OsRng` functionality (via
+    simple functions), and have `rand` depend on `os_rand` for initialisation
+3.  Have a `rand_core` sub-crate including `OsRng` functionality, the `Rng`
+    and/or `CryptoRng` trait(s), and a crypto-approved PRNG; `rand` would
+    depend on `rand_core`
+4.  Keep all generators in `rand` and move all distributions (including
+    `random()`) to a `distributions` crate (or same split but with `rng` and
+    `rand` crate names)
+5.  Split into even more crates...
+
+*Personally,* I feel that we should stick to a single `Rng` trait as above. I
+am not against splitting the implementation into multiple crates, but see
+little benefit (a split *might* facilitate review of cryptographic code).
+
+### Debug
+
+The strawman revision now specifies `pub trait Rng: Debug {...}`, i.e. requires
+all implementations to implement `Debug`. In many cases this requires only
+that implementing types are prefixed with `#[derive(Debug)]`, although in some
+cases it adds requirements on internal values.
+
+Is this useful?
+
+### Extension traits
+
+The `SeedableRng` trait should remain as is:
+
+```rust
+pub trait SeedableRng<Seed>: Rng {
+    fn reseed(&mut self, Seed);
+    fn from_seed(seed: Seed) -> Self;
+}
+```
+
+Another trait could be added to allow jump-ahead:
+
+```rust
+pub trait JumpableRng: Rng {
+    /// Return a copy of self, mutated as if `next_u32` had been called `steps`
+    /// times
+    fn jump_u32(&self, steps: usize) -> Self;
+    
+    // Also jump_u64() ? For most 64-bit generators they should be the same;
+    // for 32-bit generators, jump_u64 should jump twice as fast.
+}
+```
+
+And a trait could allow entropy injection. (Does the type `T` need to be a
+template parameter?)
+
+```rust
+pub trait InjectableRng<T>: Rng {
+    /// Add entropy to the RNG, adjusting the sequence of numbers produced.
+    fn inject(&mut self, entropy: T);
+}
+```
+
+These traits can be added in the future without breaking compatibility, however
+they may be worth discussing now.
+
+### Creation of RNGs
```

Comment:
@Lokathor I'm not against generators having associated methods that take another rng as source, if that's appropriate for that generator. The thing that I didn't like was using the `Rand` trait to seed generators and having those methods buried amongst the `Rand` implementations for a bunch of primitive values. I found that surprising.

Without `Rand` we have a few options:

- Use `SeedableRng`, but this would run into coherence issues for `impl<R: Rng>`
- Use associated methods

With associated methods, generators have more flexibility to return `Result` if they need to, and only offer constructors that make sense for them.

From the point of view of rng consumers, we offer safe and ergonomic generators in `rand` that they can use. From the point of view of rng producers, maybe we need to think about it a bit more. I think having good reference implementations with code comments geared towards anyone writing generators would be enough.

I don't think we need to explicitly capture this, but I do think it's worth exploring.